File: go/s0103/main.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"github.com/joho/godotenv"
	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
	"io"
	"log"
	"net/http"
	"os"
	"strconv"
	"strings"
)

type FinalAnswer struct {
	Task   string          `json:"task"`
	APIKey string          `json:"apikey"`
	Answer CalibrationData `json:"answer"`
}

type CalibrationData struct {
	APIKey      string     `json:"apikey"`
	Description string     `json:"description"`
	Copyright   string     `json:"copyright"`
	TestData    []TestData `json:"test-data"`
}

type TestData struct {
	Question string `json:"question"`
	Answer   int    `json:"answer"`
	Test     *Test  `json:"test,omitempty"`
}

type Test struct {
	Question string `json:"q,omitempty"`
	Answer   string `json:"a,omitempty"`
}

func main() {
	ctx := context.Background()
	err := godotenv.Load("../../.env")
	if err != nil {
		log.Fatalf("could not load env variables %v", err)
	}
	loadJsonFile, err := os.ReadFile("json.txt")
	if err != nil {
		log.Fatalln("something went wrong while loading json file", err)
	}
	var calibrationData = CalibrationData{}

	if err := json.Unmarshal(loadJsonFile, &calibrationData); err != nil {
		log.Fatalln("something went wrong while reading json data", err)
	}

	var questionsToModel []string
	for _, d := range calibrationData.TestData {
		if d.Test != nil && len(d.Test.Question) > 0 {
			questionsToModel = append(questionsToModel, d.Test.Question)
		}
	}

	if len(questionsToModel) == 0 {
		log.Fatal("questions not found")
	}

	log.Printf("there are questions to the model")
	openaiClient := openai.NewClient(option.WithAPIKey(os.Getenv("OPENAI_API_KEY")))
	response := callModelForAnswer(ctx, openaiClient, questionsToModel)

	var finalTestData []Test
	if err := json.Unmarshal([]byte(response), &finalTestData); err != nil {
		log.Fatalln("could not unmarshall model response", err)
	}
	for i := range calibrationData.TestData {
		if calibrationData.TestData[i].Test != nil && calibrationData.TestData[i].Test.Question == "" && calibrationData.TestData[i].Test.Answer == "" {
			calibrationData.TestData[i].Test = nil
		} else {
			for _, r := range finalTestData {
				if calibrationData.TestData[i].Test != nil && calibrationData.TestData[i].Test.Question == r.Question {
					calibrationData.TestData[i].Test.Answer = r.Answer
					break
				}
			}
		}
		calibrationData.TestData[i].Answer = validateCalculation(calibrationData.TestData[i].Question)
	}
	final := &FinalAnswer{
		APIKey: os.Getenv("AI_DEVS_API_KEY"),
		Task:   "JSON",
		Answer: calibrationData,
	}

	jsonString, _ := json.Marshal(final)
	if err := os.WriteFile("final.json", jsonString, os.ModePerm); err != nil {
		log.Fatalln("writing json file failed", jsonString)
	}

	client := http.Client{}
	resp, err := client.Post(fmt.Sprintf("%s/report", os.Getenv("CENTRALA_HOST")), "application/json", bytes.NewBuffer(jsonString))
	if err != nil {
		log.Fatalln("something went wrong while sending final report", err)
	}
	content, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Fatalln("could not read fina report response", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusBadRequest {
		log.Fatalf("bad request %s", string(content))
	}
	if resp.StatusCode == http.StatusOK {
		log.Println("final report sent with success")
		log.Println(string(content))
	}
}

func validateCalculation(question string) int {
	numbers := strings.Split(question, " + ")
	first, _ := strconv.Atoi(numbers[0])
	second, _ := strconv.Atoi(numbers[1])
	return first + second
}

func prepareSystemMessage() openai.ChatCompletionMessageParam {
	return openai.ChatCompletionMessageParam{
		Role: openai.F(openai.ChatCompletionMessageParamRole("system")),
		Content: openai.F[interface{}]("You will receive the list of questions. " +
			"You need to answer as short as possible, the best answer is 1 word if possible. " +
			"The response need to be written in json format. " +
			"The example of json format is presented below:  " +
			"[{\"q\":\"What is the capital city of Germany?\",\"a\":\"Berlin\"}]"),
	}
}

func prepareUserMessage(questions []string) openai.ChatCompletionMessageParam {
	stringQuestions := strings.Join(questions, "\n")
	return openai.ChatCompletionMessageParam{
		Role:    openai.F(openai.ChatCompletionMessageParamRole("user")),
		Content: openai.F[interface{}](stringQuestions),
	}
}

func callModelForAnswer(ctx context.Context, client *openai.Client, questions []string) string {
	chatCompletion, err := client.Chat.Completions.New(ctx, openai.ChatCompletionNewParams{
		Messages: openai.F([]openai.ChatCompletionMessageParamUnion{prepareSystemMessage(), prepareUserMessage(questions)}),
		Model:    openai.F(openai.ChatModelGPT4oMini),
	})

	if err != nil {
		log.Fatalln("error while calling openai", err)
	}
	answers := chatCompletion.Choices[0].Message.Content
	log.Printf("response %s", answers)
	return answers
}

File: python/s0103/main.py
import json
import os
import sys

import requests
from openai import OpenAI
from dotenv import load_dotenv
from pathlib import Path
import logging
from json import JSONEncoder
from typing import List, Dict, Any

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    handlers=[
        logging.StreamHandler(sys.stdout)
    ]
)


class Test:
    def __init__(self, question: str, answer: str):
        self.question = question
        self.answer = answer


    def to_dict(self) -> Dict[str, Any]:
        return {
            "q": self.question,
            "a": self.answer
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "Test":
        return cls(
            question=data.get("q"),
            answer=data.get("a")
        )

    def __repr__(self) -> str:
        return f"Test(question={self.question!r}, answer={self.answer!r})"



class TestData:
    def __init__(self, question: str, answer: str, test: Test):
        self.question = question
        self.answer = answer
        self.test = test

    def to_dict(self) -> Dict[str, Any]:
        return {
            "question": self.question,
            "answer": self.answer,
            "test": self.test.to_dict() if self.test else None
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "TestData":
        return cls(
            question=data["question"],
            answer=data["answer"],
            test=Test.from_dict(data.get("test")) if data.get("test") else None
        )

    def __repr__(self) -> str:
        return (
            f"TestData("
            f"question={self.question!r}, "
            f"answer={self.answer!r}, "
            f"test={self.test!r})"
        )



class CalibrationData:
    def __init__(self, apikey: str, description: str, copyright: str, test_data: list[TestData]):
        self.apikey = apikey
        self.description = description
        self.copyright = copyright
        self.test_data = test_data

    def to_dict(self) -> Dict[str, Any]:
        return {
            "apikey": self.apikey,
            "description": self.description,
            "copyright": self.copyright,
            "test-data": [td.to_dict() for td in self.test_data]
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "CalibrationData":
        return cls(
            apikey=data["apikey"],
            description=data["description"],
            copyright=data["copyright"],
            test_data=[TestData.from_dict(td) for td in data.get("test-data")]
        )

    def __repr__(self) -> str:
        return (
            f"CalibrationData("
            f"apikey={self.apikey!r}, "
            f"description={self.description!r}, "
            f"copyright={self.copyright!r}, "
            f"test_data={self.test_data!r})"
        )



class FinalAnswer:
    def __init__(self, task: str, apikey: str, answer: CalibrationData):
        self.task = task
        self.apikey = apikey
        self.answer = answer

    def to_dict(self) -> Dict[str, Any]:
        return {
            "task": self.task,
            "apikey": self.apikey,
            "answer": self.answer.to_dict() if self.answer else None
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "FinalAnswer":
        return cls(
            task=data["task"],
            apikey=data["apikey"],
            answer=CalibrationData.from_dict(data.get("answer")) if data.get("answer") else None
        )

    def to_json(self) -> str:
        return json.dumps(self.to_dict())

    @classmethod
    def from_json(cls, json_str: str) -> "FinalAnswer":
        data = json.loads(json_str)
        return cls.from_dict(data)

    def __repr__(self) -> str:
        return (
            f"FinalAnswer("
            f"task={self.task!r}, "
            f"apikey={self.apikey!r}, "
            f"answer={self.answer!r})"
        )


class FinalAnswerEncoder(JSONEncoder):
    def default(self, o):
        if hasattr(o, "to_dict"):
            return o.to_dict()
        return super().default(o)


def prepare_system_message():
    return {
        "role": "system",
        "content": "You will receive the list of questions. " +
                   "You need to answer as short as possible, the best answer is 1 word if possible. " +
                   "The response need to be written in json format. " +
                   "The example of json format is presented below:  " +
                   "[{\"q\":\"What is the capital city of Germany?\",\"a\":\"Berlin\"}]",
    }


def prepare_user_message(questions: list[str]):
    return {
        "role": "user",
        "content": "\n".join(questions)
    }


def call_model_for_answers(questions: list[str]):
    client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))

    chat_completion = client.chat.completions.create(
        messages=[prepare_system_message(), prepare_user_message(questions)],
        model="gpt-4o-mini",
    )
    answers = chat_completion.choices[0].message.content
    logging.info(f"answers are: {answers}")
    return answers


def validate_calculation(question: str) -> int:
    split = question.split(" + ")
    first = split[0]
    second = split[1]
    return int(first) + int(second)


def main():
    dotenv_path = Path('../../.env')
    load_dotenv(dotenv_path=dotenv_path)

    with open('json.txt', 'r') as file:
        file_contents = file.read()
        calibration_data = CalibrationData.from_dict(json.loads(file_contents))

    questions_to_model = []
    for data in calibration_data.test_data:
        if data.test is not None and data.test.question is not None and len(data.test.question) > 0:
            questions_to_model.append(data.test.question)

    if len(questions_to_model) == 0:
        logging.error("questions not found")
        sys.exit(1)


    logging.info("there are questions to the model")
    answers = call_model_for_answers(questions=questions_to_model)
    parsed_answers = [Test.from_dict(item) for item in json.loads(answers)]

    for test_data in calibration_data.test_data:
        for t in parsed_answers:
            if test_data.test is not None and test_data.test.question == t.question:
                test_data.test.answer = t.answer
                break

        test_data.answer = validate_calculation(test_data.question)

    final_answer = FinalAnswer(
        apikey=os.environ.get("AI_DEVS_API_KEY"),
        task="JSON",
        answer=calibration_data
    )

    final_json = json.dumps(final_answer, indent=4, cls=FinalAnswerEncoder)
    resp = requests.post(f"{os.environ.get('CENTRALA_HOST')}/report", data=final_json, headers={"Content-Type": "application/json"})

    if resp.status_code == 400:
        logging.error(f"bad request: {resp.content}")
    elif resp.status_code == 200:
        logging.info(f"final report sent with success")
        logging.info(resp.content)


if __name__ == '__main__':
    main()
